pipelinedp4j/main/com/google/privacy/differentialprivacy/pipelinedp4j/spark/SparkTable.kt

```diff
@@ -149,7 +167,40 @@ class SparkTable<K, V>(val data: Dataset<Tuple2<K, V>>,
         return filterKeys(stageName) {k -> k in allowedKeysHashSet}
     }
 
+    /**
+     * Randomly samples values per key. The output table will contain same keys as this table, each key will appear
+     * only once. The number of values per key will be at most [count].
+     * It uses window partition by function which requires an extra shuffle and sort operation and introduces
+     * an extra step to transfer data over network but is a scalable and efficient approach for large dataset.
+     */
     override fun samplePerKey(stageName: String, count: Int): SparkTable<K, List<V>> {
-        TODO("Not yet implemented")
+        val listEncoder = Encoders.kryo(List::class.java) as org.apache.spark.sql.Encoder<List<V>>
+        val withRandValueEncoder = Encoders.tuple(keyEncoder, valueEncoder, Encoders.DOUBLE())
+        val withRandomValueEncoder = Encoders.tuple(keyEncoder, valueEncoder, Encoders.INT())
+
+        // Generate a random score for each record in dataset
+        val withRandomDataset = data.map(
+            MapFunction { kv: Tuple2<K, V> -> Tuple3(kv._1, kv._2, Random.nextDouble()) },
+            withRandValueEncoder)
+
+        // Partition records by key and order them withIn each partition window by the random score and assign a sequential row_number to them
+        val windowSpec = Window.partitionBy("_1").orderBy("_3")
+        val withRowNumberDataset = withRandomDataset
```

Comment:
what does "with" mean in the name? `rowNumberDataset` seems more appropriate name for a values (same for encoders and other datasets). But of is the common way to name datasets and encoders in Spark, let's keep it.